send_budget_alert/app.py

```diff
@@ -0,0 +1,71 @@
+import json
+import synapseclient
+import re
+import os
+
+USER_ID_PATTERN = re.compile("service-catalog_(\\d+)")
+
+
+def parse_user_id_from_subject(s):
+    if s is None:
+        return None
+
+    result = USER_ID_PATTERN.search(s)
+
+    if result is None:
+        return None
+    else:
+        return result.group(1)
+
+
+def lambda_handler(event, context):
+    """Sample pure Lambda function
+
+    Parameters
+    ----------
+    event: dict, required
+        https://docs.aws.amazon.com/lambda/latest/dg/with-sns.html
+        We use three fields from each record:
+        (1) 'subject' the notification subject
+        (2) 'message' this is the notification to forward
+        (3) MessageAttributes.SynapseId.Value the Synapse user Id
+
+    context: object
+        Lambda Context runtime methods and attributes
+
+    Returns
+    ------
+    JSON map with keys:
+        errorCode: 0 if no error, non-zero otherwise
+        errorMessage: describes error if errorCode is not 0
+    """
+
+    errorCode = 0
+    errorMessage = None
+    try:
+        synapse_user_name = os.getenv("SYNAPSE_USER_NAME")
+        synapse_password = os.getenv("SYNAPSE_PASSWORD")
+        synapse_client = synapseclient.Synapse()
+        synapse_client.login(synapse_user_name, synapse_password)
+        for record in event["Records"]:
+            sns = record["Sns"]
+            subject = sns["Subject"]
+            message = sns["Message"]
+            userId = parse_user_id_from_subject(subject)
+            if userId is None:
+                raise ValueError("No user id in "+subject)
```

Comment:
What will happen with errors raised by this script beyond logging to Lambda? The consequences of not alerting a user that their budget is running out could be significant, so I wonder if there should some kind of fall through point that tell IT when a budget event is broadcast but not handled by the email-to-user function.

Reply:
My expectation is that the failures will occur while we are debugging.  Once we get it working ( once we ensure that the user id is correctly inserted into the subject line) there won't be any errors.

Having said that, it would be good to configure every headless component of the system to log errors and either configure alarms or have a dashboard that's regularly reviewed.